presto-parser/src/main/java/com/facebook/presto/sql/tree/CopyQueryVisitor.java

```diff
@@ -0,0 +1,325 @@
+    }
+    @Override
+    protected Node visitQuery(Query node, C context)
+    {
+        if (node.getLocation().isPresent()) {
+            return new Query(
+                    node.getLocation().get(),
+                    node.getWith().map(this::processSameType),
+                    processSameType(node.getQueryBody()),
+                    node.getOrderBy(),
+                    node.getOffset(),
+                    node.getLimit());
+        }
+        else {
+            return new Query(
+                    node.getWith().map(this::processSameType),
+                    processSameType(node.getQueryBody()),
+                    node.getOrderBy(),
+                    node.getOffset(),
+                    node.getLimit());
+        }
+    }
+
+    @Override
+    protected Node visitQuerySpecification(QuerySpecification node, C context)
+    {
+        if (node.getLocation().isPresent()) {
+            return new QuerySpecification(
+                    node.getLocation().get(),
+                    node.getSelect(),
+                    node.getFrom().map(this::processSameType),
+                    node.getWhere(),
+                    node.getGroupBy(),
+                    node.getHaving(),
+                    node.getOrderBy(),
+                    node.getOffset(),
+                    node.getLimit());
+        }
+        else {
+            return new QuerySpecification(
+                    node.getSelect(),
+                    node.getFrom().map(this::processSameType),
+                    node.getWhere(),
+                    node.getGroupBy(),
+                    node.getHaving(),
+                    node.getOrderBy(),
+                    node.getOffset(),
+                    node.getLimit());
+        }
+    }
+
+    @Override
+    protected Node visitUnion(Union node, C context)
+    {
+        List<Relation> newRelations = processNodeList(node.getRelations());
+        if (node.getLocation().isPresent()) {
+            return new Union(
+                    node.getLocation().get(),
+                    newRelations,
+                    node.isDistinct());
+        }
+        else {
+            return new Union(
+                    newRelations,
+                    node.isDistinct());
+        }
+    }
+
+    @Override
+    protected Node visitIntersect(Intersect node, C context)
+    {
+        List<Relation> newRelations = processNodeList(node.getRelations());
+
+        if (node.getLocation().isPresent()) {
+            return new Intersect(
+                    node.getLocation().get(),
+                    newRelations,
+                    node.isDistinct());
+        }
+        else {
+            return new Intersect(
+                    newRelations,
+                    node.isDistinct());
+        }
+    }
+
+    @Override
+    protected Node visitExcept(Except node, C context)
+    {
+        Relation left = processSameType(node.getLeft());
```

Comment:
We can do some minor optimizations like return same node if both left and right haven't changed. Same for others.